assayist/client/modify.py

```diff
@@ -0,0 +1,103 @@
+# SPDX-License-Identifier: GPL-3.0+
+
+import neomodel
+
+from assayist.common.models.source import Component
+
+
+@neomodel.db.transaction
+def set_component_names(c_name, c_type, c_namespace='', alternatives=None):
+    """
+    Create, update, and/or merge components to ensure one canonical component remains.
+
+    :param str c_name: the canonical name of the component
+    :param str c_type: the canonical type of the component
+    :param str c_namespace: the canonical namespace of the component
+    :raises ValueError: if the input isn't the proper type or c_name's value is also in alternatives
+    """
+    for arg in (c_name, c_type, c_namespace):
+        if not isinstance(arg, str):
+            raise ValueError('c_name, c_type, and c_namespace must be strings')
+
+    if alternatives is None:
+        alternatives = []
+    if not isinstance(alternatives, list) and not isinstance(alternatives, tuple):
```

Comment:
This should probably just check if it's an `abc.Iterable`. `ArrayProperty` constructs a list out of any iterable passed to it under the hood, so sets are also valid for example.

Reply:
As discussed in IRC, `collections.abc.Iterable` will be `True` when checking this on a string.

Reply:
One minor improvement might be providing the types as a tuple: `isinstance(alternatives, (tuple, list))`:
> If classinfo is a tuple of class or type objects (or recursively, other such tuples), return true if object is an instance of any of the classes or types. [[ref](https://docs.python.org/3.7/library/functions.html#isinstance)]